lib/src/signed_video_h26x_auth.c

```diff
@@ -115,6 +115,99 @@ decode_sei_data(signed_video_t *self, const uint8_t *payload, size_t payload_siz
   return status;
 }
 
+/**
+ * Computes the GOP hash from |nalu_hash_list| and compares it with the GOP hash received from the
+ * signature.
+ */
+static bool
+verify_gop_hash(signed_video_t *self)
+{
+  gop_info_t *gop_info = self->gop_info;
+  const size_t hash_size = self->verify_data->hash_size;
+  svrc_t status = SV_UNKNOWN_FAILURE;
+
+  // Compute the GOP hash from |nalu_hash_list|.
+  status = compute_partial_gop_hash(
+      self, gop_info->nalu_hash_list, gop_info->nalu_list_idx, gop_info->computed_gop_hash);
+  // Compare the computed GOP hash with the received GOP hash.
+  bool hashes_match =
+      (memcmp(gop_info->computed_gop_hash, self->received_gop_hash, hash_size) == 0);
+
+  return (hashes_match && (status == SV_OK));
+}
+
+/**
+ * Adds the hashes of NALUs used in computing the GOP hash and marks these NALUs as part of the GOP
+ * hash.
+ *
+ * This function iterates through the NALU list, identifies the NALUs that belong to the current
+ * GOP, and marks them as used in the GOP hash. It then copies the hashes of these NALUs to the
+ * |nalu_hash_list| for later verification of the GOP hash. Additionally, it handles specific cases
+ * such as moving linked hash values for I-frames and ensures that SEI-related NALUs are
+ * appropriately handled.
+ */
+void
+prepare_for_link_and_gop_hash_verification(signed_video_t *self, h26x_nalu_list_item_t *sei)
+{
+  // Ensure the `self` pointer is valid
+  assert(self);
+
+  // Initialize pointers and variables
+  h26x_nalu_list_t *nalu_list = self->nalu_list;
+  uint8_t *hash_list = self->gop_info->nalu_hash_list;
+  int *list_idx = &self->gop_info->nalu_list_idx;
+  assert(nalu_list);
+  const size_t hash_size = self->verify_data->hash_size;
+  h26x_nalu_list_item_t *item = NULL;
+  const uint8_t *hash_to_add = NULL;
+  bool found_next_gop = false;
+  bool found_item_after_sei = false;
+
+  h26x_nalu_list_print(nalu_list);
+
+  // Start with the first item in the NALU list and reset the list index.
+  item = nalu_list->first_item;
+  *list_idx = 0;
+
+  // Iterate through the NALU list until the end of the current GOP or SEI item is found.
+  while (item && !(found_next_gop || found_item_after_sei)) {
+    // Skip non-pending items
+    if (item->validation_status != 'P') {
+      item = item->next;
+      continue;
+    }
+
+    // Ensure that only non-missing NALUs (which have non-null pointers) are processed.
+    assert(item->nalu);
+
+    // Track when the current item follows the SEI or marks the beginning of the next GOP.
+    found_item_after_sei = (item->prev == sei);
+    found_next_gop = (item->nalu->is_first_nalu_in_gop && !item->need_second_verification);
+
+    // Skip GOP SEI items as they do not contribute to the GOP hash.
+    if (item->nalu->is_gop_sei) {
+      item = item->next;
+      continue;
+    }
+
+    hash_to_add = item->need_second_verification ? item->second_hash : item->hash;
+    // Copy the hash to the list if there is enough space
```

Comment:
What will happen if there is no space for the hash? Here is an example scenario:
Assume the camera generated this GOP: IPPPPIS and the nalu_list was filled up
Then someone adds one frame: IPPPPBIS

In the case where linking is done the old way you'll have these hash lists
IPPPPI vs IPPPPB
but the new way of linking will have this
IPPPP vs IPPPP